chore(claude): setup CLAUDE.md for the monorepo. Setup mcp for kona (#18824)
ethereum-optimism/optimism: CLAUDE.md

> # CLAUDE.md

Reviewer: I like the progressive disclosure section from this post:
https://www.humanlayer.dev/blog/writing-a-good-claude-md

It might be worth having a shorter root CLAUDE.md that provides brief higher-level context and each individual package has it's own CLAUDE.md with specific commands and file references to keep the context focused

Author: Good suggestion for future iteration - we already have a separate `kona/CLAUDE.md` for Kona-specific guidance. Could explore adding more package-specific files as the repo evolves.

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

## Repository Overview

> This is the Optimism monorepo containing core components of the OP Stack - the decentralized software stack that powers Optimism and forms the backbone of OP Mainnet and Base. The codebase is primarily Go with Solidity smart contracts and a Rust subproject (Kona).

Reviewer 2: I'd suggest "forms the backbone of OP Mainnet and the super chain". OP Mainnet is a bit special because we run it but we have lots of great partners running chains and shouldn't exclude them.

Author: Applied - changed to "OP Mainnet and the Superchain".

## Build Commands

### Initial Setup
```bash
# Install mise dependency manager (required)
mise trust mise.toml
mise install

# Initialize submodules
make submodules
```

### Building
```bash
# Build everything (Go + contracts)
make build

# Build only Go components
make build-go

# Build only contracts
make build-contracts

# Build specific components
make op-node
make op-batcher
make op-proposer
make op-challenger
make cannon
make op-program
```

### Linting
```bash
# Go linting
make lint-go
make lint-go-fix

# Solidity linting (from packages/contracts-bedrock)
just lint
just lint-fix
```

### Testing

**Go tests:**
```bash
# Run all Go tests
make go-tests

# Run tests with short flag (faster)
make go-tests-short

# Run tests for a specific package
go test ./op-node/...

> go test ./op-batcher/...

Reviewer 2: This is bad advice. Each package has its own make or just file with a `test` target that should be used. Running `go test` directly can fail to setup the required dependencies and cause test failures.

Author: Applied - now uses `just test` in package directories which handles dependencies properly.

# Run a single test
go test ./op-node/rollup/derive -run TestDerivation
```

**Solidity tests:**
```bash
cd packages/contracts-bedrock

# Run all contract tests
just test

# Run specific test
forge test --match-test testFunctionName

# Run tests with verbosity

> forge test -vvv

Reviewer 3: ```suggestion
# Run all contract tests with the faster 'lite' build mode enabled
just test-dev

# Run a specific test
just test-dev --match-test testFunctionName

# Run tests with verbosity
just test-dev -vvv
```

Author: Applied - now uses `just test-dev` with the suggested format.

# Run upgrade path tests (requires ETH_RPC_URL)

> just test-upgrade

Reviewer 3: ```suggestion
# Run upgrade path tests ()
FORK_BASE_CHAIN=mainnet \
FORK_OP_CHAIN=base \
ETH_RPC_URL=<rpc-url-here> \
just test-upgrade
```

Author: Applied - added the full env vars example.

```

**E2E tests:**
```bash
# Action tests (deterministic, synchronous)
cd op-e2e && go test ./actions/...

# System tests (full system integration)

> cd op-e2e && go test ./system/...

Reviewer 2: This is missing dependencies and will get failures.  They will require things like cannon prestates to be created first. I'm not sure if have dedicated make targets for these but anything we'd document here should have a build target that handles dependencies (including any introduced in the future).

Author: Applied - added `make pre-test` prerequisite documentation. The make targets handle cannon prestates and contract artifact dependencies.

Located in `packages/contracts-bedrock/src/`. Key contract categories:
- `L1/`: L1 contracts (OptimismPortal, SystemConfig, bridges)
- `L2/`: L2 predeploys (CrossDomainMessenger, L2ToL1MessagePasser)
- `cannon/`: MIPS emulator contracts for fault proofs
- `dispute/`: Dispute game contracts

### Cannon (Fault Proofs)

Cannon is an onchain MIPS instruction emulator enabling fault proofs:
- `cannon/`: MIPS emulator CLI and `mipsevm` Go implementation
- Runs `op-program` one instruction at a time for dispute resolution
- Contracts in `packages/contracts-bedrock/src/cannon/`

### Kona (Rust Components)

The `kona/` directory contains Rust implementations of OP Stack components. See `kona/CLAUDE.md` for Kona-specific guidance:
```bash
cd kona
just build-native  # Build
just test          # Test
just lint-native   # Lint
```

### Testing Infrastructure

- **op-e2e/actions/**: Deterministic state-transition tests with mock clock. Tests run synchronously without parallelism.
- **op-e2e/system/**: Full system integration tests (deprecated for new tests).
- **op-acceptance-tests/**: New acceptance test framework.
- **kurtosis-devnet/**: Kurtosis-based devnet for testing.

> ## Solidity Style Guide

Reviewer 3: Can you link it to the actual STYLE_GUIDE.md file in the repo or is that not advisable in agent docs?

Reviewer 3: This is definitely much less verbose than that doc...

Author: Added a link to the full style guide at `packages/contracts-bedrock/book/src/contributing/style-guide.md` and kept a condensed summary of key points.

When working with Solidity files:

### Documentation
- Use `///` for NatSpec comments, `//` for regular inline comments
- `@notice` for external-facing docs, `@dev` for internal notes
- Line length: 100 characters

### Naming Conventions
- Function parameters: prefix with underscore (`_amount`)
- Return arguments: suffix with underscore (`result_`)
- Immutables: `SCREAMING_SNAKE_CASE`, `internal`, with hand-written getter
- Custom errors: `ContractName_ErrorDescription`

### Testing
- Test function naming: `test_functionName_reason_succeeds/reverts`
- Test contracts: `TargetContract_FunctionName_Test`

### Upgradeability
- Contracts should extend `Initializable` and use `reinitializer(initVersion())`
- Call `_disableInitializers()` in constructor

> - All non-library contracts must implement `ISemver` with `version()`

Reviewer 3: ```suggestion
- All non-library contracts must implement `ISemver` with `version()`
- `immutable` variables are discouraged in upgradeable contracts
```

Author: Applied - added note about immutable variables being discouraged.

## Key Configuration Files

- `mise.toml`: Tool versions (Go 1.24.10, Rust 1.92.0, Foundry, etc.)
- `go.mod`: Go module definition
- `packages/contracts-bedrock/foundry.toml`: Foundry configuration
- `docker-bake.hcl`: Docker build configuration

## Development Workflow

1. PRs should target the `develop` branch for backwards-compatible changes

> 2. Contract changes in `packages/contracts-bedrock/src` may require feature branches

Reviewer 3: The feature branch decision is pretty nuanced and I think is outside the scope of this doc. But I would like to add this contract dev guidance:

```suggestion
2. Use `just pr` (in `packages/contracts-bedrock`) to run all contracts checks prior to submitting contract changes
```

Reviewer 2: Yeah please don't encourage usage of feature branches. If you need one, you're doing it wrong. Feature branches are the opposite of continuous integration (aka always be release ready).

Author: Applied - replaced feature branch guidance with `just pr` for running contract checks.
